kimchi/src/circuits/polynomials/generic.rs

```diff
@@ -1,43 +1,35 @@
-//! This module implements the double generic gate,
-//! which contains two generic gates.
-//!
-//! A generic gate is simply the 2-fan in gate specified in the
-//! vanilla PLONK protocol that allows us to do:
-//!
-//! * addition of two registers (into an output register)
-//! * or multiplication of two registers
-//! * equality of a register with a constant
-//!
-//! In each cases, registers can also be scaled by a constant.
-//!
-//! The layout of the gate is the following:
-//!
-//! |  0 |  1 |  2 |  3 |  4 |  5 | 6 | 7 | 8 | 9 | 10 | 11 | 11 | 12 | 13 | 14 |
-//! |:--:|:--:|:--:|:--:|:--:|:--:|:-:|:-:|:-:|:-:|:--:|:--:|:--:|:--:|:--:|:--:|
-//! | l1 | r1 | o1 | l2 | r2 | o2 |   |   |   |   |    |    |    |    |    |    |
-//!
-//! where l1, r1, and o1 (resp. l2, r2, o2)
-//! are the left, right, and output registers
-//! of the first (resp. second) generic gate.
-//!
-//! For the selector:
-//!
-//! |  0 |  1 |  2 |  3 |  4 |  5 | 6 | 7 | 8 | 9 | 10 | 11 | 11 | 12 | 13 | 14 |
-//! |:--:|:--:|:--:|:--:|:--:|:--:|:-:|:-:|:-:|:-:|:--:|:--:|:--:|:--:|:--:|:--:|
-//! | l1 | r1 | o1 | m1 | c1 | l2 | r2 | o2 | m2 | c2  |    |    |    |    |    |
-//!
-//! with m1 (resp. m2) the mul selector for the first (resp. second) gate,
-//! and c1 (resp. c2) the constant selector for the first (resp. second) gate.
-//!
-//! The polynomial looks like this:
-//!
-//! <pre>
-//! [
-//!   alpha1 * (w0 * coeff0 + w1 * coeff1 + w2 * coeff2 + w0 * w1 * coeff3 + coeff4) +
-//!   alpha2 * (w3 * coeff5 + w4 * coeff6 + w5 * coeff7 + w3 w4 coeff8 + coeff9)
-//! ] * generic_selector
-//! </pre>
-//!
+//! This module implements the double generic gate.
+
+//~ The double generic gate contains two generic gates.
+//~
+//~ A generic gate is simply the 2-fan in gate specified in the
+//~ vanilla PLONK protocol that allows us to do:
+//~
+//~ * addition of two registers (into an output register)
+//~ * or multiplication of two registers
+//~ * equality of a register with a constant
+//~
+//~ In each cases, registers can also be scaled by a constant.
```

Comment:
This feels a little under-specified, since in practice we use it for more than the options above. For example,
* we constrain `x = 1 or -1` with `x*x-1 = 0`,
  - we use this for the 'sign' for signed n-bit integers
* we treat an `x = 1 or -1` as if it were a boolean mask with `(x+1)*y/2 = z`, which expands to `0 * x + 2^{-1} * y + 2^{-1} * x*y + (-1) * z + 0 = 0`
  - this allows us to compute `if x > 0 then x else 0` for a signed n-bit integer `x`.

Really, this gives us a way to define a gate that satisfies a linear function `f(x, y, x*y, z) = 0`. I don't know if there's a more elegant way to state this though; perhaps these additional examples are sufficient.

Reply:
yeah that's true, perhaps these use-cases could be added to the `GenericGateSpec` as well

Reply:
I don't think we need explicit constructors for them, but a constructor like
```rust
    FullGeneric {
        /// Optional coefficient that can be multiplied with the left operand.
        left_coeff: Option<F>,
        /// Optional coefficient that can be multiplied with the right operand.
        right_coeff: Option<F>,
        /// Optional coefficient that can be multiplied with the multiplication result.
        mul_coeff: Option<F>,
        /// Optional coefficient that can be multiplied with the output.
        output_coeff: Option<F>,
        /// Optional constant to add to the expression.
        constant_term: Option<F>,
    },
```
might be nice for completeness.